handlers/utils.go

```diff
@@ -22,13 +22,59 @@ const defaultPatchType = "merge"
 
 var defaultPromClientSupplier = prometheus.NewClient
 
-func checkNamespaceAccess(ctx context.Context, nsServ business.NamespaceService, namespace string) (*models.Namespace, error) {
-	return nsServ.GetNamespace(ctx, namespace)
+func checkNamespaceAccess(ctx context.Context, nsServ business.NamespaceService, namespace string, cluster string) (*models.Namespace, error) {
+	return nsServ.GetNamespaceByCluster(ctx, namespace, cluster)
 }
 
-func createMetricsServiceForNamespace(w http.ResponseWriter, r *http.Request, promSupplier promClientSupplier, namespace string) (*business.MetricsService, *models.Namespace) {
-	metrics, infoMap := createMetricsServiceForNamespaces(w, r, promSupplier, []string{namespace})
-	if result, ok := infoMap[namespace]; ok {
+// Create Metrics Service for a single namespace and no cluster parameter. It will check permissions for all the clusters, and return the service
+// if it has permissions at least one of the clusters
+func createMetricsServiceForNamespaceMC(w http.ResponseWriter, r *http.Request, promSupplier promClientSupplier, ns models.Namespace) (*business.MetricsService, *models.Namespace) {
+	layer, err := getBusiness(r)
+	if err != nil {
+		RespondWithError(w, http.StatusInternalServerError, err.Error())
+		return nil, nil
+	}
+	prom, err := promSupplier()
+	if err != nil {
+		log.Error(err)
+		RespondWithError(w, http.StatusServiceUnavailable, "Prometheus client error: "+err.Error())
+		return nil, nil
+	}
+	var nsInfo nsInfoError
+
+	nsLists, err := layer.Namespace.GetNamespaceClusters(r.Context(), ns.Name)
+	if err != nil || len(nsLists) == 0 {
+		log.Infof("No clusters found.")
+		// TWe cant find the cluster list, we asume it is local
+		nsLists = make([]models.Namespace, 1)
+		nsLists[0] = models.Namespace{Cluster: config.Get().KubernetesConfig.ClusterName, Name: ns.Name}
+	}
+
+	for _, nsCluster := range nsLists {
+		nsNoErrors := len(nsLists)
+		info, err := checkNamespaceAccess(r.Context(), layer.Namespace, nsCluster.Name, nsCluster.Cluster)
+		nsInfo = nsInfoError{info: info, err: err}
+		if err != nil {
+			nsNoErrors--
+		}
+		// At least, one ns with permissions
+		if nsNoErrors > 0 {
+			nsInfo = nsInfoError{info: info, err: nil}
+		}
+	}
+	metrics := business.NewMetricsService(prom)
+	if nsInfo.err != nil {
+		RespondWithError(w, http.StatusForbidden, "Cannot access namespace data: "+nsInfo.err.Error())
+		return nil, nil
+	}
+
+	return metrics, nsInfo.info
```

Comment:
For callers of this function, I don't think it's deterministic which namespace you'll get? Should this return a list of namespaces now instead of a single namespace? The namespace model is scoped to a specific cluster rather than spanning across all clusters even though that's how they're currently represented in the frontend.

Reply:
Yes, you made a good point! 

I guess the question here would be, from the namespace info, the creation timestamp will be used, which one could we used for the metrics query? The newest/oldest? This would be for the query start time, https://github.com/kiali/kiali/blob/master/handlers/metrics.go#L300

Reply:
I will ping @jshaughn in case he has some ideas here

Reply:
I agree, a non-deterministic return should be avoided but I'm not sure a list makes sense here.  The caller only needs one metrics service.  The way metrics queries work, correct me if I'm wrong, it will either be against a single cluster (i.e. the request includes the `cluster` param), or across all clusters.   I think the caller is probably responsible for knowing what it's going to do.  And so I think this function probably should take a cluster argument.  If the namespace exists on the cluster and the user has access then return the metricsService.  If the cluster arg is "" then assume it's a cross-cluster metrics query and make sure the user has access to the namespace on all its clusters.  If so, return the metricsService.

As for the creation timestamp, that's sort of an odd one.  It's used to mainly to make sure we're not querying farther back than makes sense.  So for a single cluster query it should be set to the namespace on that cluster.  For all clusters ("") its hould be set to the oldest create time of the namespace clusters.

Reply:
Thanks for clarifying. I have changed it to return the oldest one, and return an error if there are not permissions in one of the namespaces. (Maybe we should just get into account the namespaces that are coming from the clusters that the user has permissions from?)

Regarding to use the cluster argument, there is a separate function that has the cluster argument and returns the metrics service. I've thought it would be more clear to have a separate function as it can have different logic to just handle this specific situation. Maybe it makes more sense to return a single metrics service but a list of namespace info?

Reply:
I was going to add the list o namespaces to the extractIstioMetricsQueryParams but, there are other methods that are using it and they just have a single namespace. So, in order to not to change them, I have created an aux method to have just this functionality.